compiler/rustc_infer/src/infer/error_reporting/mod.rs

```diff
@@ -1623,6 +1669,53 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
         self.note_error_origin(diag, cause, exp_found);
     }
 
+    pub fn get_impl_future_output_ty(&self, ty: Ty<'tcx>) -> Option<Ty<'tcx>> {
```

Comment:
https://github.com/rust-lang/rust/pull/76765/files#diff-b5c71ca5f71c77ba51507d5ba5cfe4afa15578aaa8f2f0385ef88423cb69744fR88 has a similar fn but it operates on `rustc_*hir*::Ty<'_>` instead of this, which I think is `rustc_middle::Ty<'_>`

Should they be merged?  In that PR, we specifically need to check the HIR because we need to know if its an async fn, but after that we could switch to using `rustc_middle` (can you get an `rustc_middle::Ty` with a local def id?)

That fn also just checks the bound against the symbol `Output`, this seems to find the future trait def and compare against that, that may be better (but I am not sure if `rustc_hir::Ty` makes that easy)

Reply:
Whenever you have a `ty::Ty<'tcx>` it is easier to do it this way, when you have access to the `hir::Ty` that comes from the return type of the `hir` `fn` return type it easier to use that directly.

We *might* be able to unify these two methods, but I've found in the past that I've had to have the same logic twice, one for the `hir` and another for `ty`, so I wouldn't be *too* torn about the "duplication".

Reply:
perhaps we can merge these separately and then if I find time I can see how hard it is to merge!

Reply:
That sounds good.